src/sidebar/components/ShareDialog/ExportAnnotations.tsx

```diff
@@ -133,69 +137,138 @@ function ExportAnnotations({
 
-  if (!exportReady) {
-    return <LoadingSpinner />;
-  }
-
-  const exportAnnotations = (e: Event) => {
-    e.preventDefault();
-
-    try {
-      const format = exportFormat.value;
+  const buildExportContent = useCallback(
+    // Can't use an arrow function with generics in a tsx file
+    // eslint-disable-next-line prefer-arrow-callback
+    function <
+      T extends ExportFormat['value'],
+      R = T extends 'json' ? JSONExportContent : string,
+    >(format: T): R {
       const annotationsToExport =
         selectedUser?.annotations ?? exportableAnnotations;
-      const filename = `${customFilename ?? defaultFilename}.${format}`;
-
       switch (format) {
         case 'json': {
-          const exportData = annotationsExporter.buildJSONExportContent(
+          return annotationsExporter.buildJSONExportContent(
             annotationsToExport,
             { profile },
-          );
-          downloadJSONFile(exportData, filename);
-          break;
+          ) as R;
         }
         case 'txt': {
-          const exportData = annotationsExporter.buildTextExportContent(
+          return annotationsExporter.buildTextExportContent(
             annotationsToExport,
             {
               groupName: group?.name,
               defaultAuthority,
               displayNamesEnabled,
             },
-          );
-          downloadTextFile(exportData, filename);
-          break;
+          ) as R;
         }
         case 'csv': {
-          const exportData = annotationsExporter.buildCSVExportContent(
+          return annotationsExporter.buildCSVExportContent(
             annotationsToExport,
             {
               groupName: group?.name,
               defaultAuthority,
               displayNamesEnabled,
             },
-          );
-          downloadCSVFile(exportData, filename);
-          break;
+          ) as R;
         }
         case 'html': {
-          const exportData = annotationsExporter.buildHTMLExportContent(
+          return annotationsExporter.buildHTMLExportContent(
             annotationsToExport,
             {
               groupName: group?.name,
               defaultAuthority,
               displayNamesEnabled,
             },
-          );
-          downloadHTMLFile(exportData, filename);
-          break;
+          ) as R;
+        }
+        default:
+          throw new Error('Invalid format');
+      }
+    },
+    [
+      annotationsExporter,
+      defaultAuthority,
+      displayNamesEnabled,
+      exportableAnnotations,
+      group?.name,
+      profile,
+      selectedUser?.annotations,
+    ],
+  );
+  const exportAnnotations = useCallback(
+    (e: Event) => {
+      e.preventDefault();
+
+      try {
+        const format = exportFormat.value;
+        const filename = `${customFilename ?? defaultFilename}.${format}`;
+
+        switch (format) {
+          case 'json': {
+            const exportData = buildExportContent(format);
```

Comment:
This line is duplicated on every `switch` case because the type inference doesn't work otherwise.